doc/design/requirements.md

```diff
@@ -0,0 +1,146 @@
+* Uninstall TiUp
+
+### Deployment and managing a deployed cluster
+
+TODO
+
+Better error messages when setting up and managing a cluster
+
+### Sources
+
+Also called mirrors and servers
+
+* PingCAP will host a publicly available server
+* Any third party may host a public or private server
+  - Users must opt-in to using a source
+  - Sources are strictly ordered for a given user (and can be changed by the user)
+  - Sources cannot be specified per-component
+  - All sources are checked in turn for a component
+  - If updating a component causes it to be installed from a different source, the user should be prompted
+* The user may create a local server, that server is created empty
+* Components can be cloned from any source to a local source, either a specific version, all versions, or a range of versions
+* TiUp can manage a server (possibly with a non-standard component); users should not have to manually edit any part of a server
+* Hosts of servers are responsible for any authentication, TiUp will not keep servers private but will provide API hooks for hosts to do so
+
+### Supported products
+
+* TiDB
+* TiKV (with and without TiDB)
+* TiFlash
+* Binlog
+* Blackbox probe
+* Prometheus and Grafana
+
+TiUp should be extensible with user-provided components.
+
+### Interfaces
+
+This section is getting beyond requirements and should be removed, but might be useful for deriving requirements
+
+* SQL client
+* playground
+* cluster
+
+### Non-functional requirements
+
+* Security - we must be able to securely download files and manage clusters since the DB is likely to be security-critical for users. Download binaries is likely to be over the internet. Cluster management might include cross-data center communication on a public network. No need to protect data at rest.
+* Reliability - could bugs in cluster management cause severe issues? That might be important. I think delaying update is not critical as long as it is not for too long, perhaps <1hr?
+* Performance - performance is not critical, but the CLI should be responsive (low latency). Since most tasks are expected to be long-running, there is not much pressure for high performance.
+* Memory/disk usage - unlikely to be a problem. We should provide mechanisms for automatic and manual recovery of disk space, since in the long-term it might become an issue.
+* Ease of contribution and development velocity - should be prioritised to help OSS community building. 
+* Scale - should support our largest users, i.e., 1000s of nodes in a cluster and nodes in different data centers.
+* Backwards compatibility - TODO how much do we have to preserve workflows and scripts between versions?
+
+### Non-goals
+
+Things we intend not to do, either for now or ever.
+
+* Designed to be primarily used by humans, secondarily by scripts. Not intended as an API.
+* No GUI.
+* Project is specific to TiDB and TiKV, not a generic database or software management tool.
```

Comment:
I'm not sure about this. If the user provides a component that can manage MySQL or other software, it can manage these Softwares. FYI @lonng

Reply:
That sounds fine. The way I would frame it is that if this works we don't try and stop it, but for now it is not a goal. Does that sound ok?